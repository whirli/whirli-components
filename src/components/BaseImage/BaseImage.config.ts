// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseImage/BaseImage.config';

// Package config
import { PropKeys } from './BaseImage.constants';
import type { PropImage } from './BaseImage.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  [PropKeys.IMAGE]: {
    required: true,
    type: Object as PropType<PropImage>,
  },
  ...config.props,
};
