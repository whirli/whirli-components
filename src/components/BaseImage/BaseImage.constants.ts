export const PropKeys: Record<string, string> = {
  IMAGE: 'image',
};

export const PropValues: Record<string, string[]> = {};
