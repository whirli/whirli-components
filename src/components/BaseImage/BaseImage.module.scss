@import '../../styles/Breakpoints.scss';

// default styles
.image__wrapper {
  position: relative;
  height: 0;
  width: 100%;
}

.image {
  position: absolute;
  top: 0;
  left: 0;
  object-fit: contain;
  width: 100%;
  height: 100%;
  transition: opacity ease-in-out 0.5s;
  opacity: 0;
}

.image[loaded='true'] {
  opacity: 1;
}

// breakpoint styles
$packageBreakpointStyles: ();