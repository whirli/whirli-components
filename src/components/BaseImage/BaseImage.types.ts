export interface PropImage {
  url: string;
  width: string;
  height: string;
  title: string;
}

export interface ImageProps {
  image: PropImage;
}
