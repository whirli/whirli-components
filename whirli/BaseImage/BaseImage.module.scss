@import '@whirli-components/src/components/BaseImage/BaseImage.module.scss';

// Standard styles

// Breakpoint styles
$localBreakpointStyles: ();

@include createBreakpointClasses(map-merge($packageBreakpointStyles, $localBreakpointStyles));