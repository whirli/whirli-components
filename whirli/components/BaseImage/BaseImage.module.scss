@import '@whirli-components/components/BaseImage/BaseImage.module.scss';

// Standard styles

// Breakpoint styles
$localBreakpointStyles: ();

@include createBreakpointClasses(map-merge($packageBreakpointStyles, $localBreakpointStyles));