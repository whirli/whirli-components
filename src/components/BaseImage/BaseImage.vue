<template>
  <div :class="wrapperClasses" :style="`padding-top: ${height}`">
    <picture>
      <img
        :class="imageClasses"
        :src="props.image.url"
        :alt="props.image.title"
        :height="props.image.height"
        :width="props.image.width"
        ref="defaultImage"
        loading="lazy"
      />
    </picture>
  </div>
</template>

<script setup lang="ts">
// Vue
import { ref, computed, onMounted } from '@composition';

// Types
import type { ComputedRef, Ref } from '@composition';
import type { PropImage } from './BaseImage.types';

// Constants
import { PropKeys } from './BaseImage.constants';

// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseImage/BaseImage.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseImage.config';

const ComponentStyles = ConfigStyles;

// @ts-ignore
const props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const wrapperClasses = [styles.image__wrapper, ...makeClasses(ComponentStyles, ConfigProps, props, styles)];
const imageClasses = [styles.image];

// Component specific

const defaultImage = ref(null) as Ref<HTMLElement | null>;

const height: ComputedRef<string> = computed(() => {
  const percentage =
    (parseInt((props[PropKeys.IMAGE] as unknown as PropImage).height) /
      parseInt((props[PropKeys.IMAGE] as unknown as PropImage).width)) *
    100;
  return `${percentage}%`;
});

const loadImage: () => void = (): void => {
  (defaultImage.value as unknown as HTMLElement).addEventListener('load', () => {
    (defaultImage.value as unknown as HTMLElement).setAttribute('loaded', 'true');
  });
};

onMounted((): void => {
  loadImage();
});
</script>
