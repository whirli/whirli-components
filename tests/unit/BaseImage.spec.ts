import { shallowMount } from '@vue/test-utils';
import BaseImage from '@/components/BaseImage/BaseImage.vue';

let wrapper: any;

describe('BaseImage.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseImage, {
      props: {
        image: {
          url: 'my-image.jpg',
          width: '400',
          height: '400',
          title: 'Hello world',
        },
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct src', async () => {
    expect(wrapper.find('img').attributes('src')).toBe('my-image.jpg');
  });

  it('renders the correct alt', async () => {
    expect(wrapper.find('img').attributes('alt')).toBe('Hello world');
  });

  it('returns the correct ratio', () => {
    expect(wrapper.element.style.getPropertyValue('padding-top')).toBe('100%');
  });
});
