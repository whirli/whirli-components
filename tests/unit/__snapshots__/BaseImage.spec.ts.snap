// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseImage.vue - Single classes renders the component 1`] = `
<div
  class=""
  style="padding-top: 100%;"
>
  <picture>
    <img
      alt="Hello world"
      class=""
      height="400"
      loading="lazy"
      src="my-image.jpg"
      width="400"
    />
  </picture>
</div>
`;
